File: AlgorithmLcr/Try3-Component types with known topology and corresponding positions/topofit_id/graph.py
"""Multigraph representation and exact structure reductions (DESIGN.md sec.3).

The DUT is a multigraph: nodes are junctions (0 and 1 are the port
terminals), edges are components.  Parallel edges between the same node
pair are legal and may carry different kinds.  Self loops carry zero
voltage hence zero current and are dropped.

Working at topology level only (kinds known, values unknown), four classes
of *exact* reductions remove edges/parameters that cannot influence Z:

  F1  self loops, edges outside the port-connected component, dangling
      edges (leaf at a non-port node carries no current);
  F2  same-kind parallel merge  (R: G=G1+G2, C: C=C1+C2; L is NOT mergeable
      in parallel -- two series-RL in parallel are second order);
  F3  same-kind series merge at degree-2 internal nodes (R: sum,
      C: reciprocal sum, L: L and Rd both sum);
  F4  series absorb of an R edge into an adjacent L edge (R + (L+Rd)
      = L + (Rd+R): the resistor only shifts the inductor's DCR).

Every rule preserves Z(s) exactly for ALL positive element values, so the
reduction is value-independent and can run before fitting.  Merged edges
keep an aggregation expression tree over the original edges; the tree is
evaluated on true values for benchmarks and reported to the user because
the members of a merged group are structurally unidentifiable
individually (only the aggregate is determined by Z).
"""

from __future__ import annotations

from dataclasses import dataclass

KINDS = ("R", "C", "L")

# value tuples carried by an edge/group: ("R", r) | ("C", c) | ("L", l, rd)
Val = tuple


class PortOpenError(ValueError):
    """The graph leaves the port (nodes 0-1) open at every frequency."""


def _pair(u: int, v: int) -> tuple:
    return (u, v) if u <= v else (v, u)


# ---------------------------------------------------------------------------
# aggregation expressions over original edges
# ---------------------------------------------------------------------------

def expr_leaf(idx: int, kind: str) -> tuple:
    return ("e", idx, kind)


def _agg(tag: str, children: list, kind: str) -> tuple:
    return (tag, tuple(children), kind)


def eval_group(expr: tuple, orig_vals: dict) -> tuple:
    """Evaluate an aggregation tree on original-edge values.

    ("e", i, k)        -> orig_vals[i]
    ("ser", xs, "R")   -> ("R", sum r)
    ("ser", xs, "C")   -> ("C", 1 / sum(1/c))
    ("ser", xs, "L")   -> ("L", sum l, sum of R-children r + sum of L rd)
    ("par", xs, "R")   -> ("R", 1 / sum(1/r))
    ("par", xs, "C")   -> ("C", sum c)
    """
    tag = expr[0]
    if tag == "e":
        return orig_vals[expr[1]]
    children, kind = expr[1], expr[2]
    vals = [eval_group(ch, orig_vals) for ch in children]
    if tag == "par":
        if kind == "R":
            return ("R", 1.0 / sum(1.0 / v[1] for v in vals))
        return ("C", sum(v[1] for v in vals))
    # series
    if kind == "R":
        return ("R", sum(v[1] for v in vals))
    if kind == "C":
        return ("C", 1.0 / sum(1.0 / v[1] for v in vals))
    l_tot = sum(v[1] for v in vals if v[0] == "L")
    rd_tot = sum(v[2] for v in vals if v[0] == "L") + \
        sum(v[1] for v in vals if v[0] == "R")
    return ("L", l_tot, rd_tot)


# ---------------------------------------------------------------------------
# reduction
# ---------------------------------------------------------------------------

@dataclass
class ReducedEdge:
    u: int                      # original node labels
    v: int
    kind: str                   # resulting kind
    expr: tuple                 # aggregation tree over original edges
    members: tuple              # original edge indices in the group


@dataclass
class ReductionResult:
    edges: list
    dropped: dict               # original edge idx -> reason
    n_passes: int

    @property
    def n_groups(self) -> int:
        return len(self.edges)

    def group_values(self, orig_vals: dict) -> list:
        return [eval_group(e.expr, orig_vals) for e in self.edges]

    def describe(self) -> str:
        parts = []
        for g, e in enumerate(self.edges):
            parts.append("g{}:{}[{}-{}]{{{}}}".format(
                g, e.kind, e.u, e.v, ",".join(map(str, e.members))))
        for i, why in sorted(self.dropped.items()):
            parts.append("e{}:dropped({})".format(i, why))
        return " ".join(parts)


@dataclass
class _WEdge:
    u: int
    v: int
    kind: str
    expr: tuple
    members: tuple


def _components(nodes: set, wedges: list) -> dict:
    parent = {n: n for n in nodes}

    def find(a: int) -> int:
        while parent[a] != a:
            parent[a] = parent[parent[a]]
            a = parent[a]
        return a

    for e in wedges:
        ra, rb = find(e.u), find(e.v)
        if ra != rb:
            parent[ra] = rb
    return {n: find(n) for n in nodes}


def _drop_dangling(wedges: list, dropped: dict, nodes: set):
    """Iteratively strip edges hanging off non-port leaf nodes."""
    changed_any = False
    while True:
        deg = {n: 0 for n in nodes}
        for e in wedges:
            deg[e.u] += 1
            deg[e.v] += 1
        victim = None
        for n, d in sorted(deg.items()):
            if n not in (0, 1) and d == 1:
                victim_edge = next(e for e in wedges if e.u == n or e.v == n)
                victim = (n, victim_edge)
                break
        if victim is None:
            return wedges, changed_any
        n, edge = victim
        wedges = [e for e in wedges if e is not edge]
        nodes = {m for m in nodes if m != n}
        dropped[edge.members[0]] = "dangling"
        changed_any = True


def reduce_graph(edges: list) -> ReductionResult:
    """Topology-level exact reduction to fixpoint (DESIGN.md sec.3.3)."""
    if not edges:
        raise PortOpenError("empty edge list")
    for (u, v, k) in edges:
        if k not in KINDS:
            raise ValueError("kind must be one of {}, got {!r}".format(KINDS, k))

    work = [_WEdge(u, v, k, expr_leaf(i, k), (i,))
            for i, (u, v, k) in enumerate(edges)]
    dropped = {}
    n_passes = 0
    changed = True
    while changed:
        changed = False
        n_passes += 1

        # -- F1a: self loops ------------------------------------------------
        keep = []
        for e in work:
            if e.u == e.v:
                dropped[e.members[0]] = "self-loop"
                changed = True
            else:
                keep.append(e)
        work = keep

        # -- F1b: connectivity of the port ----------------------------------
        nodes = {n for e in work for n in (e.u, e.v)} | {0, 1}
        comp = _components(nodes, work)
        root0 = comp[0]
        if comp[1] != root0:
            raise PortOpenError("nodes 0 and 1 are not connected")
        keep = []
        for e in work:
            if comp[e.u] == root0:
                keep.append(e)
            else:
                dropped[e.members[0]] = "disconnected"
                changed = True
        work = keep

        # -- F1c: dangling branches -----------------------------------------
        nodes = {n for e in work for n in (e.u, e.v)} | {0, 1}
        work, ch = _drop_dangling(work, dropped, nodes)
        changed = changed or ch

        # -- F2: same-kind parallel merges (R, C only) ----------------------
        groups = {}
        for e in work:
            groups.setdefault(_pair(e.u, e.v), []).append(e)
        merged = []
        for pair, es in groups.items():
            by_kind = {}
            for e in es:
                by_kind.setdefault(e.kind, []).append(e)
            for kind, same in by_kind.items():
                if kind in ("R", "C") and len(same) > 1:
                    merged.append(_WEdge(pair[0], pair[1], kind,
                                         _agg("par", [e.expr for e in same], kind),
                                         tuple(m for e in same for m in e.members)))
                    changed = True
                else:
                    merged.extend(same)
        work = merged

        # -- F3/F4: series merges at degree-2 internal nodes -----------------
        deg = {}
        for e in work:
            for n in (e.u, e.v):
                deg[n] = deg.get(n, 0) + 1
        series_node = None
        for n, d in sorted(deg.items()):
            if n not in (0, 1) and d == 2:
                inc = [e for e in work if e.u == n or e.v == n]
                if len(inc) == 2:
                    e1, e2 = inc
                    a = e1.v if e1.u == n else e1.u
                    b = e2.v if e2.u == n else e2.u
                    if a != b:
                        ks = {e1.kind, e2.kind}
                        if ks in ({"R"}, {"C"}, {"L"}, {"R", "L"}):
                            series_node = (n, e1, e2, a, b)
                            break
        if series_node is not None:
            n, e1, e2, a, b = series_node
            kind = "L" if "L" in {e1.kind, e2.kind} else e1.kind
            work = [e for e in work if e is not e1 and e is not e2]
            work.append(_WEdge(a, b, kind,
                               _agg("ser", [e1.expr, e2.expr], kind),
                               e1.members + e2.members))
            changed = True

    if not work:
        raise PortOpenError("no edge influences the port after reduction")
    return ReductionResult(edges=[ReducedEdge(e.u, e.v, e.kind, e.expr, e.members)
                                  for e in work],
                           dropped=dropped, n_passes=n_passes)

File: AlgorithmLcr/Try3-Component types with known topology and corresponding positions/cppversion/src/graph.cpp
#include "graph.hpp"

#include <algorithm>
#include <cstdio>
#include <numeric>
#include <set>

namespace tf {

namespace {

Expr makeLeaf(int idx, char kind) {
    Expr e;
    e.leaf = true;
    e.edgeIdx = idx;
    e.kind = kind;
    return e;
}

Expr makeAgg(char tag, std::vector<Expr> children, char kind) {
    Expr e;
    e.leaf = false;
    e.tag = tag;
    e.kind = kind;
    e.children = std::move(children);
    return e;
}

struct WEdge {
    int u = 0, v = 0;
    char kind = 'R';
    Expr expr;
    std::vector<int> members;
};

// union-find over the nodes touched by the working edges (+ ports)
std::map<int, int> componentsOf(const std::set<int>& nodes, const std::vector<WEdge>& wedges) {
    std::map<int, int> parent;
    for (int n : nodes) parent[n] = n;
    auto find = [&parent](int a) {
        while (parent[a] != a) {
            parent[a] = parent[parent[a]];
            a = parent[a];
        }
        return a;
    };
    for (const auto& e : wedges) {
        int ra = find(e.u), rb = find(e.v);
        if (ra != rb) parent[ra] = rb;
    }
    std::map<int, int> comp;
    for (int n : nodes) comp[n] = find(n);
    return comp;
}

}  // namespace

Value evalGroup(const Expr& expr, const std::vector<Value>& origVals) {
    if (expr.leaf) return origVals[(size_t)expr.edgeIdx];
    std::vector<Value> vals;
    vals.reserve(expr.children.size());
    for (const auto& ch : expr.children) vals.push_back(evalGroup(ch, origVals));
    const char kind = expr.kind;
    if (expr.tag == 'p') {  // parallel
        if (kind == 'R') {
            double g = 0.0;
            for (const auto& val : vals) g += 1.0 / val.v1;
            return Value{'R', 1.0 / g, 0.0};
        }
        double c = 0.0;
        for (const auto& val : vals) c += val.v1;
        return Value{'C', c, 0.0};
    }
    // series
    if (kind == 'R') {
        double r = 0.0;
        for (const auto& val : vals) r += val.v1;
        return Value{'R', r, 0.0};
    }
    if (kind == 'C') {
        double s = 0.0;
        for (const auto& val : vals) s += 1.0 / val.v1;
        return Value{'C', 1.0 / s, 0.0};
    }
    double lTot = 0.0, rdTot = 0.0;
    for (const auto& val : vals) {
        if (val.kind == 'L') {
            lTot += val.v1;
            rdTot += val.v2;
        } else if (val.kind == 'R') {
            rdTot += val.v1;
        }
    }
    return Value{'L', lTot, rdTot};
}

std::vector<Value> ReductionResult::groupValues(
    const std::vector<Value>& origVals) const {
    std::vector<Value> out;
    out.reserve(edges.size());
    for (const auto& e : edges) out.push_back(evalGroup(e.expr, origVals));
    return out;
}

std::string ReductionResult::describe() const {
    std::string out;
    for (size_t g = 0; g < edges.size(); ++g) {
        const auto& e = edges[g];
        char buf[96];
        std::string mem;
        for (size_t q = 0; q < e.members.size(); ++q) {
            if (q) mem += ",";
            mem += std::to_string(e.members[q]);
        }
        std::snprintf(buf, sizeof(buf), "g%zu:%c[%d-%d]{%s} ", g, e.kind, e.u, e.v,
                      mem.c_str());
        out += buf;
    }
    for (const auto& [i, why] : dropped) {
        out += "e" + std::to_string(i) + ":dropped(" + why + ") ";
    }
    if (!out.empty()) out.pop_back();
    return out;
}

namespace {

void dropDangling(std::vector<WEdge>& work, std::map<int, std::string>& dropped,
                  std::set<int>& nodes, bool& changedAny) {
    while (true) {
        std::map<int, int> deg;
        for (int n : nodes) deg[n] = 0;
        for (const auto& e : work) {
            deg[e.u] += 1;
            deg[e.v] += 1;
        }
        int victimNode = -1;
        const WEdge* victimEdge = nullptr;
        for (const auto& [n, d] : deg) {  // ascending node label
            if (n != 0 && n != 1 && d == 1) {
                for (const auto& e : work)
                    if (e.u == n || e.v == n) {
                        victimNode = n;
                        victimEdge = &e;
                        break;
                    }
                break;  // only the FIRST such node is considered per pass
            }
        }
        if (victimNode < 0) return;
        // copy the victim's data first: removing it from  frees the
        // buffer the pointer refers into
        std::vector<int> victimMembers = victimEdge->members;
        std::vector<WEdge> next;
        bool removed = false;
        for (const auto& e : work) {
            if (!removed && &e == victimEdge) {
                removed = true;
                continue;
            }
            next.push_back(e);
        }
        work = std::move(next);
        nodes.erase(victimNode);
        for (int m : victimMembers) dropped[m] = "dangling";
        changedAny = true;
    }
}

}  // namespace

ReductionResult reduceGraph(const std::vector<std::tuple<int, int, char>>& edges) {
    if (edges.empty()) throw PortOpenError("empty edge list");
    for (const auto& [u, v, k] : edges) {
        (void)u;
        (void)v;
        if (k != 'R' && k != 'C' && k != 'L')
            throw std::invalid_argument(std::string("kind must be one of R|L|C, got '") +
                                        k + "'");
    }
    std::vector<WEdge> work;
    for (size_t i = 0; i < edges.size(); ++i) {
        auto [u, v, k] = edges[i];
        WEdge e;
        e.u = u;
        e.v = v;
        e.kind = k;
        e.expr = makeLeaf((int)i, k);
        e.members = {(int)i};
        work.push_back(std::move(e));
    }
    std::map<int, std::string> dropped;
    int nPasses = 0;
    bool changed = true;
    while (changed) {
        changed = false;
        ++nPasses;

        // -- F1a: self loops ----------------------------------------------
        {
            std::vector<WEdge> keep;
            for (const auto& e : work) {
                if (e.u == e.v) {
                    for (int m : e.members) dropped[m] = "self-loop";
                    changed = true;
                } else {
                    keep.push_back(e);
                }
            }
            work = std::move(keep);
        }

        // -- F1b: connectivity of the port --------------------------------
        {
            std::set<int> nodes;
            for (const auto& e : work) {
                nodes.insert(e.u);
                nodes.insert(e.v);
            }
            nodes.insert(0);
            nodes.insert(1);
            std::map<int, int> comp = componentsOf(nodes, work);
            int root0 = comp[0];
            if (comp[1] != root0)
                throw PortOpenError("nodes 0 and 1 are not connected");
            std::vector<WEdge> keep;
            for (const auto& e : work) {
                if (comp[e.u] == root0) {
                    keep.push_back(e);
                } else {
                    for (int m : e.members) dropped[m] = "disconnected";
                    changed = true;
                }
            }
            work = std::move(keep);
        }

        // -- F1c: dangling branches ----------------------------------------
        {
            std::set<int> nodes;
            for (const auto& e : work) {
                nodes.insert(e.u);
                nodes.insert(e.v);
            }
            nodes.insert(0);
            nodes.insert(1);
            bool ch = false;
            dropDangling(work, dropped, nodes, ch);
            changed = changed || ch;
        }

        // -- F2: same-kind parallel merges (R, C only) ---------------------
        {
            // groups keyed by normalized pair, first-encounter order
            std::vector<std::pair<std::pair<int, int>, std::vector<const WEdge*>>> groups;
            std::map<std::pair<int, int>, int> groupIdx;
            for (const auto& e : work) {
                auto pr = e.u <= e.v ? std::make_pair(e.u, e.v) : std::make_pair(e.v, e.u);
                auto it = groupIdx.find(pr);
                if (it == groupIdx.end()) {
                    groupIdx[pr] = (int)groups.size();
                    groups.push_back({pr, {&e}});
                } else {
                    groups[(size_t)it->second].second.push_back(&e);
                }
            }
            std::vector<WEdge> merged;
            for (const auto& [pr, es] : groups) {
                // by-kind buckets in first-encounter order
                std::vector<std::pair<char, std::vector<const WEdge*>>> byKind;
                std::map<char, int> kindIdx;
                for (const WEdge* e : es) {
                    auto it = kindIdx.find(e->kind);
                    if (it == kindIdx.end()) {
                        kindIdx[e->kind] = (int)byKind.size();
                        byKind.push_back({e->kind, {e}});
                    } else {
                        byKind[(size_t)it->second].second.push_back(e);
                    }
                }
                for (const auto& [kind, same] : byKind) {
                    if ((kind == 'R' || kind == 'C') && same.size() > 1) {
                        std::vector<Expr> children;
                        std::vector<int> members;
                        for (const WEdge* e : same) {
                            children.push_back(e->expr);
                            members.insert(members.end(), e->members.begin(),
                                           e->members.end());
                        }
                        WEdge ne;
                        ne.u = pr.first;
                        ne.v = pr.second;
                        ne.kind = kind;
                        ne.expr = makeAgg('p', std::move(children), kind);
                        ne.members = std::move(members);
                        merged.push_back(std::move(ne));
                        changed = true;
                    } else {
                        for (const WEdge* e : same) merged.push_back(*e);
                    }
                }
            }
            work = std::move(merged);
        }

        // -- F3/F4: series merge at the first eligible degree-2 node -------
        {
            std::map<int, int> deg;
            for (const auto& e : work) {
                deg[e.u] += 1;
                deg[e.v] += 1;
            }
            const WEdge *e1 = nullptr, *e2 = nullptr;
            int aOut = 0, bOut = 0;
            bool found = false;
            for (const auto& [n, d] : deg) {  // ascending label
                if (n == 0 || n == 1 || d != 2) continue;
                std::vector<const WEdge*> inc;
                for (const auto& e : work)
                    if (e.u == n || e.v == n) inc.push_back(&e);
                if (inc.size() != 2) continue;
                e1 = inc[0];
                e2 = inc[1];
                int a = e1->u == n ? e1->v : e1->u;
                int b = e2->u == n ? e2->v : e2->u;
                if (a == b) continue;
                char k1 = e1->kind, k2 = e2->kind;
                bool mergeable = (k1 == k2 && (k1 == 'R' || k1 == 'C' || k1 == 'L')) ||
                                 ((k1 == 'R' && k2 == 'L') || (k1 == 'L' && k2 == 'R'));
                if (mergeable) {
                    aOut = a;
                    bOut = b;
                    found = true;
                    break;
                }
            }
            if (found) {
                char kind = (e1->kind == 'L' || e2->kind == 'L') ? 'L' : e1->kind;
                // copy both edges' payloads first: reassigning  frees
                // the buffer e1/e2 point into
                Expr e1Expr = e1->expr, e2Expr = e2->expr;
                std::vector<int> members = e1->members;
                members.insert(members.end(), e2->members.begin(), e2->members.end());
                std::vector<WEdge> next;
                for (const auto& e : work)
                    if (&e != e1 && &e != e2) next.push_back(e);
                work = std::move(next);
                WEdge ne;
                ne.u = aOut;
                ne.v = bOut;
                ne.kind = kind;
                ne.expr = makeAgg('s', {e1Expr, e2Expr}, kind);
                ne.members = std::move(members);
                work.push_back(std::move(ne));
                changed = true;
            }
        }
    }

    if (work.empty()) throw PortOpenError("no edge influences the port after reduction");
    ReductionResult out;
    for (const auto& e : work) {
        ReducedEdge re;
        re.u = e.u;
        re.v = e.v;
        re.kind = e.kind;
        re.expr = e.expr;
        re.members = e.members;
        out.edges.push_back(std::move(re));
    }
    out.dropped = std::move(dropped);
    out.nPasses = nPasses;
    return out;
}

}  // namespace tf
